WIP #15613 - Add DHCPv6 support to proxy

xprazak2 opened the pull request:
Initial attempt for DHCPv6 on proxy.

TODO:
- [x] Only one Libvirt network supported - check if it is v4 or v6 before suggesting ip
- [ ] Ad file observers for libvirt
- [ ] Check for reserved v6 ips (multicast)
- [ ] ISC
- [ ] native MS
- [ ] tests

Commit:
Fixes #15613 - [WIP] Add DHCPv6 support to proxy

dmitri-d commented:
@xprazak2: We should chat about this when you have a moment.

dLobatog commented:
@witlessbird Let us know what comes out of that chat here :smile: - I'm interested in the implementation and would be happy to help with coding & testing as I'm already familiar with how this works in Foreman.

cc-ing @timogoebel too

dmitri-d commented:
@xprazak2, @dLobatog, @timogoebel: There are a few things I would like to discuss:
- the interface with backends will need to be updated to deal with ipv6, which is a good opportunity to bring it up to date with the latest changes (https://github.com/theforeman/smart-proxy/pull/409 and https://github.com/witlessbird/smart-proxy/tree/use-dhcps-api)
- The first PR above needs to be merged before this work, as it significantly changes our interaction with isc dhcpd
- I expect significant changes (please see https://github.com/witlessbird/smart-proxy/tree/use-dhcps-api if you are curious) the to ms dhcp provider are going to be introduced early next week, this is going to affect this work as well

timogoebel commented on lib/proxy/validations.rb:
```diff
@@ -14,8 +16,17 @@ def valid_mac? mac
-    raise Error, "Invalid IP Address #{ip}" unless ip =~ /(\d{1,3}\.){3}\d{1,3}/
+    IPAddr.new(ip)
     ip
+  rescue IPAddr::InvalidAddressError
```
The exception thrown differs between ruby versions, please add something like this: https://github.com/theforeman/foreman/blob/79c0664a008afe0fea54a0b5256426f3867a02bc/app/models/subnet.rb#L288-L292

timogoebel commented:
@xprazak2 : I just had a brief look at the code. Thanks for taking over this issue. Please keep me posted with your progress. I'd be happy to help.

I currently paused my work on the IPv6 stuff to wait for the EFI PRs from @lzap . I think they are a requirement to test orchestration as a whole.

I agree, that we should merge the PR from @witlessbird before this. Let's try to speed this up a little. :-)

timogoebel commented:
@xprazak2 : #409 hat just been merged. I think, you can go ahead with this. Curious to see an update.

xprazak2 commented:
This is what I have so far: I refactored libvirt provider and cleaned up libvirt_main, there are no file observers yet. I made changes to isc only to find out omshell lacks support for IPv6. I asked on isc-users list if anyone knows about plans to add it but I haven't received any replies yet. In the meantime, I will move on to ms provider.

timogoebel commented:
@xprazak2 : Just fyi, I started with the Foreman parts: https://github.com/timogoebel/foreman/tree/15612-ipv6-dhcpv6
This is still very wip.

dmitri-d commented on lib/proxy/validations.rb:
```diff
@@ -14,8 +17,26 @@ def valid_mac? mac
   # validates the ip address
   def validate_ip ip
-    raise Error, "Invalid IP Address #{ip}" unless ip =~ /(\d{1,3}\.){3}\d{1,3}/
+    IPAddr.new(ip)
```
Please don't rely on implicit validation during object instantiation. Instead:

```
raise "Error" if (ip =~ =~ Resolv::AddressRegex).nil?
```

dmitri-d commented on lib/proxy/validations.rb:
```diff
@@ -14,8 +17,26 @@ def valid_mac? mac
+  end
+
+  # we may not want to raise error in some cases bu skip the ip instead
+  def soft_validate_ip ip
```
That is a very confusing name. Can rely on "!" do signal presence of side-effects in calls. Then `validate_ip` becomes `validate_ip!` and `soft_validate_ip` becomes `validate_ip`.

dmitri-d replied:
Also, not need for logging here.

dmitri-d commented on lib/proxy/validations.rb:
```diff
@@ -24,6 +45,16 @@ def validate_mac mac
     mac.downcase
   end
 
+  def validate_mac_or_uid mac, uid
```
Please break this up into two methods

dmitri-d commented on lib/proxy/validations.rb:
```diff
@@ -39,4 +70,11 @@ def validate_record record
     record
   end
 
+  private
```
No need to use 'private' anywhere other than in libraries. All it does is here is making it harder to test things.

dmitri-d commented on modules/dhcp/dhcp_api.rb:
```diff
@@ -31,7 +31,7 @@ def load_subnet_data
     begin
       content_type :json
-      server.subnets.map{|s| {:network => s.network, :netmask => s.netmask, :options => s.options}}.to_json
+      server.subnets.map{|s| {:network => s.network, :netmask => s.netmask, :options => s.options, :prefix => s.prefix}}.to_json
```
I think the api controller needs:
- change the interface to providers -- drop `load_subnet` and `load_subnet_data` calls. If you look at the recent and pending changes you'll notice that those aren't being used. In other words, one call to provider per operation.
- keep ipv4 and ipv6 calls separate (by possibly prefixing the latter with 'ipv6'?)
  - You may have broken current (and older) versions of foreman with this implementation of `get /`
  - create/delete/free ip calls are going to be different for ipv4 and ipv6, better to keep them separate to avoid guessing which needs to be called
  - Similarly for record retrieval: At least dhcp option calls are going to be ipv4/ipv6-specific, probably easier to keep those separate

dmitri-d commented on modules/dhcp_common/record.rb:
```diff
@@ -5,15 +5,15 @@ module Proxy::DHCP
   # represent a DHCP Record
   class Record
```
I think the record class can be removed, it's mostly being used as a container for options hash. Additionally: 
- subnet should be stored as a string: it's super annoying having to construct a Subnet object every time a lease or resevation is constructed, only so the subnet address string can be used later).
- Ipv4 and ipv6 can be kept separate, would also allow for meaningful initialization (no set_mac_or_uid-like methods)
  - validation should be moved out of constructor (if required at all -- we only create these objects from data received from isc/ms/etc backends)

dmitri-d commented on modules/dhcp_common/subnet.rb:
```diff
@@ -48,15 +47,27 @@ def include? ip
 
-    def cidr
-      IPAddr.new(netmask).to_i.to_s(2).count("1")
+    def prefix
```
Please remove these:
- internal code not meant to be extended
- you will get an "undefined method" error anyway

dmitri-d commented on modules/dhcp_common/subnet.rb:
```diff
@@ -10,16 +10,15 @@
     include Proxy::Validations
 
-    def initialize network, netmask, options = {}
+    def initialize network, options = {}
```
I think this class can be removed too in favour of ipv4 and ipv6-specific ones:
- unused_ip code should be moved out from Subnet -- at least two providers have their own implementation. It's going to be ipv4/ipv6 specific too.
- Everywhere else subnet class is either used as a container for options hash (which is broken) or for its ip address/mask

dmitri-d commented on modules/dhcp_isc/configuration_loader.rb:
```diff
@@ -48,11 +32,55 @@ def load_dependency_injection_wirings(container, settings)
+      require 'dhcp_common/subnet/ipv6'
+    end
+
+    private
```
no need for private here

dmitri-d commented on modules/dhcp_isc/configuration_loader.rb:
```diff
@@ -48,11 +32,55 @@ def load_dependency_injection_wirings(container, settings)
+    end
+
+    def load_wirings_for_files(container, settings)
+      versions.each do |ver|
```
`[4, 6].each do |ver|`

dmitri-d commented on modules/dhcp_isc/dhcp_isc_plugin.rb:
```diff
@@ -2,16 +2,19 @@ module ::Proxy::DHCP::ISC
 
     requires :dhcp, ::Proxy::VERSION
-    validate_readable :config, :leases
+    validate_readable :config4, :leases4, :config6, :leases6
```
What if only ipv4 or ipv6 is enabled/used?

dmitri-d commented on modules/dhcp_isc/isc_file_parser.rb:
```diff
@@ -7,10 +7,11 @@ module Proxy::DHCP::ISC
   class FileParser
```
As mentioned above, I think it's better to keep ipv6 and ipv4 data structures separate. Because of this big chunks of code will move from this class into ipv4/ipv6 specific ones, which might reduce it to a module?

dmitri-d commented on modules/dhcp_libvirt/configuration_loader.rb:
```diff
@@ -2,16 +2,22 @@ module Proxy::DHCP::Libvirt
     def load_dependency_injection_wirings(container, settings)
       container.dependency :memory_store, ::Proxy::MemoryStore
-      container.dependency :subnet_service, (lambda do
+      container.singleton_dependency :subnet_service, (lambda do
```
why?

dmitri-d commented:
To summarise: 
- Now is a good opportunity to change the interface between providers and the api controller, we should do it. Otherwise there will be way more work after this.
- ipv4 and ipv6 operations and data structures can be splt from each other: this will make implementation and testing easier
- I would suggest to add/update tests early, esp. around various parsers

dmitri-d commented:
> I made changes to isc only to find out omshell lacks support for IPv6. I asked on isc-users list if anyone knows about plans to add it but I haven't received any replies yet. 

I can't find much (only looked very briefly) on omapi support for ipv6. Looks like isc is developing a new dhcp server http://kea.isc.org/wiki, perhaps it has better ipv6 support? 

> In the meantime, I will move on to ms provider.

Please use changes in https://github.com/theforeman/smart-proxy/pull/445, not the 'develop' branch.

dmitri-d commented:
I finally had a chance to play with omshell talking to a dhcpd6 instance. It appears to mostly work ok  (haven't tried setting options on host object, but that bit should work). 

The main disconnect I found is between ipv4 host ids and ipv6 duids (host-identifier option vs. option dhcp-client-identifier). While it is possible to set/modify the former, it is impossible to do the same for ipv6 duids, which restricts how forman-managed hosts can be deployed (see https://tools.ietf.org/html/rfc3315#section-11 for more information).

That aside, here are some examples. Locating a record via hardware address:

```
> new host
obj: host
> set hardware-address = 01:00:80:a2:55:67:34
obj: host
hardware-address = 01:00:80:a2:55:67:34
> open
obj: host
hardware-address = 01:00:80:a2:55:67:34
ip-address = 20:01:0d:b8:00:01:00:01:00:00:00:00:00:00:07:01
dhcp-client-identifier = 00:01:00:01:14:ed:66:c1:08:00:27:94:08:40
name = "test"
hardware-type = 00:00:00:01
> close
obj: <null>
```

Creating a new host record:

```
> new host
obj: host
> set name = "test-another"
obj: host
name = "test-another"
> set ip-address = 20:01:0d:b8:00:01:00:01:00:00:00:00:00:00:07:02
obj: host
name = "test-another"
ip-address = 20:01:0d:b8:00:01:00:01:00:00:00:00:00:00:07:02
> set hardware-address = 01:00:80:a2:55:67:35
obj: host
name = "test-another"
ip-address = 20:01:0d:b8:00:01:00:01:00:00:00:00:00:00:07:02
hardware-address = 01:00:80:a2:55:67:35
> create
obj: host
name = "test-another"
ip-address = 20:01:0d:b8:00:01:00:01:00:00:00:00:00:00:07:02
hardware-address = 01:00:80:a2:55:67:35
hardware-type = 00:00:00:00
> close
obj: <null>
> new host
obj: host
> set name = "test-another"
obj: host
name = "test-another"
> open
obj: host
name = "test-another"
ip-address = 20:01:0d:b8:00:01:00:01:00:00:00:00:00:00:07:02
hardware-address = 01:00:80:a2:55:67:35
hardware-type = 00:00:00:00
```

Host record removal:

```
obj: host
name = "test-another"
ip-address = 20:01:0d:b8:00:01:00:01:00:00:00:00:00:00:07:02
hardware-address = 01:00:80:a2:55:67:35
hardware-type = 00:00:00:00
> remove
obj: <null>
> new host
obj: host
> set hardware-address = 01:00:80:a2:55:67:35
obj: host
hardware-address = 01:00:80:a2:55:67:35
> open
can't open object: not found
obj: host
hardware-address = 01:00:80:a2:55:67:35
```

dmitri-d commented:
> I can't find much (only looked very briefly) on omapi support for ipv6. Looks like isc is developing a new dhcp server http://kea.isc.org/wiki, perhaps it has better ipv6 support?

kea doesn't have external api other than for server status and restarts. Unless we are prepared to invest time into adding this functionality it's not  an option, at least for the time being.

lzap commented:
@witlessbird did you get any updates on the ISC DHCP mailing list in this regard?

dmitri-d commented:
@lzap see my comment above.

dmitri-d commented:
I don't think we reached an impasse here -- most of stuff is doable, it just needs doing. Removing the label.

timogoebel commented:
What's the status here, guys? @xprazak2, are you still working on this?

dmitri-d commented:
@timogoebel: I started working on this.

xprazak2 commented:
@witlessbird, I am sorry I could not finish this in timely manner, but priorities shifted 😞. If you want to take over, I have no objections.

dmitri-d commented:
I'm closing this PR -- I think ipv6 support will require a different approach to the one we currently use.
